This simplifies the number parsing routine and removes our dependency on strtod_l

lemire opened the pull request:
This trims out a lot of "complicated" code in the number parsing routine. The main idea is to do less work to avoid doing other work, and rather to do all of the work eagerly. Putting it differently, this is a more "branchless" approach. We also reorganize the lookup tables and slightly reduce their size.  The results appear to be somewhat positive, mostly neutral.

We have the general problem that not all platforms provide a *sane* (or at least, an accurate) strtod, and since we want a locale-independent approach, we have to rely on strtod_l which is non-standard (although well supported). Still, some strtod_l appear not to be precise (just like strtod). Out there, you will find Gay's C code from the early 2000s, but it has not aged well. There are libraries like abseil, but they do not support all the platforms we want to support.  This here follows a proposal by Nigel Tao. It relies on slow but relatively understandable code credited to Ken Thompson. What we care about, in simdjson, is not speed, since this code will almost never be called, but rather that it is correct and easy to debug. So I have adapted and tested Nigel's code to a C++ style that is more "simdjson-like", trimming what was unneeded. The fallback code is non throwing and non allocating.

We also add quite a bit of new testing to make sure that the number parsing is exact.

Fixes https://github.com/simdjson/simdjson/issues/1190

AMD Rome, gcc 9:

|filename | ref. speed (GB/s) |candidate speed | difference |
|:--------| ----------:| --------------:| ----------:|
|numbers.json|0.893|0.932|4.4%|
|gsoc-2018.json|3.278|3.375|3.0%|
|repeat.json|2.347|2.386|1.7%|
|mesh.json|0.793|0.805|1.5%|
|canada.json|0.782|0.793|1.5%|
|marine_ik.json|0.801|0.810|1.1%|
|mesh.pretty.json|1.310|1.323|1.0%|
|semanticscholar-corpus.json|2.064|2.084|1.0%|
|apache_builds.json|2.480|2.501|0.8%|
|tree-pretty.json|2.174|2.190|0.8%|
|google_maps_api_response.json|2.113|2.127|0.7%|
|citm_catalog.json|2.634|2.650|0.6%|
|update-center.json|1.913|1.924|0.6%|
|twitter_api_response.json|2.479|2.492|0.5%|
|random.json|1.532|1.539|0.5%|
|google_maps_api_compact_response.json|1.094|1.095|0.1%|
|twitter_timeline.json|1.656|1.655|-0.0%|
|twitterescaped.json|1.206|1.205|-0.1%|
|github_events.json|2.749|2.745|-0.1%|
|instruments.json|2.078|2.075|-0.2%|
|twitter.json|2.406|2.395|-0.4%|
|twitter_api_compact_response.json|2.011|1.999|-0.6%|

clang 9 (AMD Rome)


|filename | ref. speed (GB/s) |candidate speed | difference |
|:--------| ----------:| --------------:| ----------:|
|canada.json|0.644|0.710|10.3%|
|mesh.pretty.json|1.130|1.168|3.4%|
|repeat.json|2.352|2.396|1.9%|
|apache_builds.json|2.293|2.306|0.5%|
|github_events.json|2.485|2.492|0.3%|
|citm_catalog.json|2.515|2.521|0.3%|
|random.json|1.445|1.448|0.2%|
|google_maps_api_compact_response.json|0.978|0.977|-0.1%|
|google_maps_api_response.json|1.953|1.952|-0.1%|
|semanticscholar-corpus.json|2.047|2.043|-0.2%|
|gsoc-2018.json|3.424|3.402|-0.6%|
|twitter_timeline.json|1.453|1.443|-0.7%|
|update-center.json|1.817|1.802|-0.9%|
|instruments.json|1.846|1.823|-1.3%|
|twitterescaped.json|1.169|1.152|-1.4%|
|tree-pretty.json|1.998|1.958|-2.0%|
|twitter.json|2.249|2.203|-2.1%|
|mesh.json|0.702|0.685|-2.3%|
|numbers.json|0.811|0.789|-2.7%|
|marine_ik.json|0.729|0.708|-2.8%|
|twitter_api_response.json|2.304|2.232|-3.1%|
|twitter_api_compact_response.json|1.845|1.761|-4.5%|

jkeiser commented on src/jsoncharutils_tables.h:
```diff
@@ -201,8 +201,8 @@ const uint32_t digit_to_val32[886] = {
 
-constexpr int FASTFLOAT_SMALLEST_POWER = -344;
-constexpr int FASTFLOAT_LARGEST_POWER = 308;
+constexpr int simdjson_smallest_power = -344;
```
I'm actually OK with renaming all constants to lowercase--is that what you usually do?

I'm finding it looks a LOT better.

lemire replied:
"I'm actually OK with renaming all constants to lowercase--is that what you usually do?"

=> My expectation is that these were macros before that became constants...

jkeiser commented on src/jsoncharutils_tables.h:
```diff
@@ -201,8 +201,8 @@ const uint32_t digit_to_val32[886] = {
-constexpr int FASTFLOAT_SMALLEST_POWER = -344;
-constexpr int FASTFLOAT_LARGEST_POWER = 308;
+constexpr int simdjson_smallest_power = -344;
+constexpr int simdjson_largest_power = 308;
```
Also, these don't need to be prefixed with simdjson_ because they are already in the simdjson namespace. Wonderful benefit of using constants instead of macros :)

lemire replied:
You are correct, they should not be prefixed.

lemire commented:
@jkeiser My expectation is that this PR should be performance neutral. I need to figure out why it seems to have negative effects. It is possible that it is random stage 2 variations... but I'd like to be sure.

The goal is to simplify the code.

lemire commented:
Note that this is deliberately being left in draft mode.

lemire commented:
Ok. I think I have made the performance effect closer to neutral by adding a branch at the right location.

lemire commented:
This has been converted to a draft PR. I want to add more tests before merging this in, to make sure that bugs are not introduced.

lemire commented:
PR https://github.com/simdjson/simdjson/pull/1195 adds more tests and should be merged first.

lemire commented:
Rebased again dlemire/issue1190

jkeiser commented:
OK, this is next up on my TODO. Been pretty fully booked bulldozing the On Demand stuff.

lemire commented:
This is not the final step. I have other simplifications upcoming. Ultimately, we can make it so that it is super improbable that you will need the fallback. But getting rid of the fallback requires me to prove that everything is correct, which takes time.

lemire commented:
(By not the final step, I mean that there will be more PRs not that this PR is incomplete.)

jkeiser commented on src/generic/stage2/numberparsing.h:
```diff
@@ -110,61 +102,57 @@ simdjson_really_inline bool compute_float_64(int64_t power, uint64_t i, bool neg
   i <<= lz;
+
+
+  // We are going to need to do some 64-bit arithmetic to get a  precise product.
```
Two spaces between a and precise, not that it matters.

lemire replied:
Fixed.

jkeiser commented on src/generic/stage2/numberparsing.h:
```diff
@@ -110,61 +102,57 @@ simdjson_really_inline bool compute_float_64(int64_t power, uint64_t i, bool neg
-    // This does happen, e.g. with 7.3177701707893310e+15.
-    if (((product_middle + 1 == 0) && ((product_high & 0x1FF) == 0x1FF) &&
-         (product_low + i < product_low))) { // let us be prudent and bail out.
+  const uint32_t index = 2 * uint32_t(power - smallest_power);
```
I'm sure the compiler will (probably) optimize it, but it's worth noting here that Intel addressing modes can handle something like (small power of 2) * X + Y + Constant all in a single address lookup. It's not necessary to hyperoptimize just yet, but since the constant is applied *before* the multiplication instead of after, and this calculation is indirect, it's worth a note to check the assembly for `power_of_five_128[index]` and `power_of_five_128[index+1]` to see if they are single instructions.

lemire replied:
@jkeiser I do not remember the details, but there are definitively cases on some processors where loads from simple addressing schemes are faster. I *think* that Intel processors can issue two loads per cycle and one store, if and only if the addressing is simple. 

Even if that is wrong, Travis reminds us that there is definitively a latency cost on intel processors...
https://stackoverflow.com/a/51664696


I'll add a remark, however, so that we can revisit this later.

jkeiser commented on src/generic/stage2/numberparsing.h:
```diff
@@ -110,61 +102,57 @@ simdjson_really_inline bool compute_float_64(int64_t power, uint64_t i, bool neg
+  // exact and no further work is needed. Having 55 bits is necessary because
+  // we need 53 bits for the mantissa but we have to have one rounding bit and
+  // we can waste a bit if the most significant bit of the product is zero.
+  if((firstproduct.high & 0x1FF) == 0x1FF) {
```
I take it `full_multiplication()` is designed such that the high bit is always 1 if the input's high bit is 1? It wasn't spelled out explicitly earlier that the output would look like this (nor on full_multiplication), and that property seems pretty critical here.

lemire replied:
No. full_multiplication is just that, it is just a * b, getting the full product. 

The high bit thing is something else.

lemire replied:
Added this remark:

```
  // The full_multiplication function computes the 128-bit product of two 64-bit words
  // with a returned value of type value128 with a "low component" corresponding to the
  // 64-bit least significant bits of the product and with a "high component" corresponding
  // to the 64-bit most significant bits of the product.
```

lemire replied:
Another remark:

```
  // Both i and power_of_five_128[index] have their most significant bit set to 1 which
  // implies that the either the most or the second most significant bit of the product 
  // is 1. We pack values in this manner for efficiency reasons: it maximizes the use
  // we make of the product.
```

jkeiser commented on src/generic/stage2/numberparsing.h:
```diff
@@ -110,61 +102,57 @@ simdjson_really_inline bool compute_float_64(int64_t power, uint64_t i, bool neg
+  // we need 53 bits for the mantissa but we have to have one rounding bit and
+  // we can waste a bit if the most significant bit of the product is zero.
+  if((firstproduct.high & 0x1FF) == 0x1FF) {
+    value128 secondproduct = full_multiplication(i, power_of_five_128[index + 1]);
```
So my intuition, looking at this line, is that we're taking i*5^(power+1) and i*5^power and then doing something to mix them, but I'm unclear precisely what. Is `firstproduct.low` the lower 64 bits of i*5^power? And `secondproduct.high` is the higher 64 bits of i*5^(power+1)? I can't figure out what it means to add those two things.

lemire replied:
I'll add a remark. Your intuition is wrong.

There is only so much that I can explain in the code but I will explain better.

lemire replied:
Adding the following...

```
    // We want to compute i * 5^q, but only care about the top 55 bits at most.
    // Consider the scenario where q>=0. Then 5^q may not fit in 64-bits. Doing
    // the full computation is wasteful. So we do what is called a "truncated
    // multiplication".
    // We take the most significant 64-bits, and we put them in 
    // power_of_five_128[index]. Usually, that's good enough to approximate i * 5^q
    // to the desired approximation using one multiplication. Sometimes it does not suffice. 
    // Then we store the next most significant 64 bits in power_of_five_128[index + 1], and
    // then we get a better approximation to i * 5^q. In very rare cases, even that
    // will not suffice, though it is seemingly very hard to find such a scenario.
    // 
    // That's for when q>=0. The logic for q<0 is somewhat similar but it is somewhat
    // more complicated.
    //
    // There is an extra layer of complexity in that we need more than 55 bits of 
    // accuracy in the round-to-even scenario.
    //
```

I am working on a full write-up.

jkeiser commented on src/generic/stage2/numberparsing.h:
```diff
@@ -179,26 +167,12 @@ simdjson_really_inline bool compute_float_64(int64_t power, uint64_t i, bool neg
   // floating-point values.
-  if (simdjson_unlikely((lower == 0) && ((upper & 0x1FF) == 0) &&
+  // We have 5**27 < 2**64 and it is the largest power of 5 to do so.
+  if (simdjson_unlikely((lower == 0) && (power >= -27) && (power <= 27)  &&
```
The inference here is that any number >= 2^64 cannot possibly be between two values because it's got too many significant digits to do so. I'm curious why that's not 2^54?

Also, double space before the && -> single space, unless this was on purpose?

lemire replied:
> The inference here is that any number >= 2^64 cannot possibly be between two values because it's got too many significant digits to do so. I'm curious why that's not 2^54?

This part has been explained much better now...

```
  // 
  // The round-to-even cases take the form of a number 2m+1 which is in (2^53,2^54]
  // times a power of two. That is, it is right between a number with binary significand
  // m and another number with binary significand m+1; and it must be the case
  // that it cannot be represented by a float itself.
  //
  // We must have that w * 10 ^q == (2m+1) * 2^p for some power of two 2^p.
  // Recall that 10^q = 5^q * 2^q.
  // When q >= 0, we must have that (2m+1) is divible by 5^q, so 5^q <= 2^54. We have that
  //  5^23 <=  2^54 and it is the last power of five to qualify, so q <= 23.
  // When q<0, we have  w  >=  (2m+1) x 5^{-q}.  We must have that w<2^{64} so
  // (2m+1) x 5^{-q} < 2^{64}. We have that 2m+1>2^{53}. Hence, we must have 
  // 2^{53} x 5^{-q} < 2^{64}.
  // Hence we have 5^{-q} < 2^{11}$ or q>= -4$  (64-bit case). 
  //
  // We require lower <= 1 and not lower == 0 because we could not prove that 
  // that lower == 0 is implied; but we could prove that lower <= 1 is a sufficient test.
```

I still need to document this better, but there is only so much I can put in comments.

lemire replied:
I have removed the double space.

jkeiser commented on src/generic/stage2/numberparsing.h:
```diff
@@ -179,26 +167,12 @@ simdjson_really_inline bool compute_float_64(int64_t power, uint64_t i, bool neg
-    // almost always should round up.
-    return false;
+    if((mantissa  << (upperbit + 64 - 53 - 2)) ==  upper) {
+      mantissa ^= 1;             // flip it so that we do not round up
```
I think this would be clearer as mantissa &= ~1 ... using ^ implies we think it might be 0 or 1 to start with, but we know it's 1.

lemire replied:
Done.

jkeiser commented on src/generic/stage2/numberparsing.h:
```diff
@@ -224,39 +198,16 @@ simdjson_really_inline bool compute_float_64(int64_t power, uint64_t i, bool neg
-  // will flat out throw an exception.
-  //
-  if ((endptr == (const char *)ptr) || (!std::isfinite(*outDouble))) {
+// We call a fallback floating-point parser that might be slow. Note
```
Ah, so you were able to remove if (endptr == ptr) because we already know there is at least one digit, due to having tried to parse it the normal way! Nice.

lemire replied:
Adding a remark. Yes.

jkeiser commented on src/generic/stage2/numberparsing.h:
```diff
@@ -398,7 +349,9 @@ simdjson_really_inline error_code write_float(const uint8_t *const src, bool neg
   // digit count is off by 1 because of the decimal (assuming there was one).
```
I'd remove this line since you explain it better down below.

lemire replied:
Agreed.

lemire replied:
Done.

jkeiser commented on src/generic/stage2/numberparsing.h:
```diff
@@ -418,7 +371,7 @@ simdjson_really_inline error_code write_float(const uint8_t *const src, bool neg
   // way we've tried: https://github.com/simdjson/simdjson/pull/990#discussion_r448497331
   // To future reader: we'd love if someone found a better way, or at least could explain this result!
-  if (simdjson_unlikely(exponent < FASTFLOAT_SMALLEST_POWER) || (exponent > FASTFLOAT_LARGEST_POWER)) {
+  if (simdjson_unlikely(exponent < smallest_power) || (exponent > largest_power)) {
```
It's worth talking about casing and constants for style in the future. We don't have to resolve it here, though.

I find it very useful for constants to look different from variables. Variables make my brain go "OK I need to track this down, see where it comes from and range of values it can have." Knowing something is a constant often allows me to ignore it and focus on more important things.

jkeiser commented on src/jsoncharutils_tables.h:
```diff
@@ -201,8 +201,8 @@ const uint32_t digit_to_val32[886] = {
 
-constexpr int FASTFLOAT_SMALLEST_POWER = -325;
-constexpr int FASTFLOAT_LARGEST_POWER = 308;
+constexpr int smallest_power = -344;
```
I'm curious what leads to the choice of -344? Mind you, I don't know why we chose -325 before either :)

lemire replied:
I have adjusted the minimal power to be -342. Powers smaller than -342 would automatically lead to zero values.

```
/**
 * The smallest non-zero float (binary64) is 2^−1074.
 * We take as input numbers of the form w x 10^q where w < 2^64.
 * We have that w * 10^-343  <  2^(64-344) 5^-343 < 2^-1076.
 * However, we have that 
 * (2^64-1) * 10^-342 =  (2^64-1) * 2^-342 * 5^-342 > 2^−1074.
 * Thus it is possible for a number of the form w * 10^-342 where 
 * w is a 64-bit value to be a non-zero floating-point number.
 *********
 * Any number of form w * 10^309 where w>= 1 is going to be 
 * infinite in binary64 so we never need to worry about powers
 * of 5 greater than 308.
 */
```

jkeiser commented on tests/random_string_number_tests.cpp:
```diff
@@ -0,0 +1,157 @@
+
+int main() {
+  // We test 1,000,000 random strings.
+  if (tester(1234344, 1000000)) {
```
It would be a nice project to just test every single one of them, probably across multiple machines or over a long time period--then we'd know this will never give us a random failure. But I'm good with this, it's good extra coverage and I don't think there's a huge chance of such a thing.

jkeiser replied:
It occurs to me that there are an infinite number of potential strings, even if there is a finite number of strings we'd probably need to test (the ones on and adjacent to the border between floats). Again, this is just a comment and seriously don't do it right now :)

lemire replied:
I am puzzled about these comments. Yes, there are definitively infinitely many strings. I test a million of them. It obviously does not establish that there is no bug. I do this to give us *confidence* that there is no bug.

It is super hard to prove that there is no possible bug. There might be. I just try to make it as unlikely as possible.

Even just checking all 64-bit floats is impossible. At least impossible given my ressources. But even if you took all 64-bit values and serialized them and checked that you got back to original, you have to contend with the fact that there an infinite number of ways to serialize a given 64-bit value.

Isn't it just life as a programmer?

jkeiser replied:
Your test is good :) I'm mostly musing. I know Rust has claimed to have done a comprehensive test of float parsing with at least one of each possible float; a quick search didn't turn it up though, I'll keep looking.

lemire replied:
Suppose you spent 100 cycles on a 4 GHz machine for each one of the possible 2**64 binary64 value, it would take you 15,000 years. But, of course, you need more than one hundred cycles per float...  Still, 15,000 processors is within reach at places like Microsoft, Google and the NSA. But that would still give you only one test per float... and you would have kept busy 15,000 processors for a year. At that scale, just ensuring that the system does what you think it does is going to be a massive engineer achievement.

At the other hand of the spectrum, 1,000,000 is not much but it is easy enough to crank this up and let it run for days.

lemire replied:
I will change the code so that one can run it as long as one wants (but the 1,000,000 is a good default).

lemire replied:
Done. So the -m 32132312 would run 32132312 tests.

lemire replied:
Finally done.

jkeiser commented on tests/basictests.cpp:
```diff
@@ -194,14 +221,14 @@ namespace number_tests {
       size_t n = snprintf(buf, sizeof(buf), "1e%d", i);
       if (n >= sizeof(buf)) { abort(); }
-      fflush(NULL);
+      // fflush(NULL); // This should hopefully not be needed.
```
Yeah, I imagine you can just remove it.

lemire replied:
Removing.

jkeiser commented on src/jsoncharutils_tables.h:
```diff
@@ -215,652 +215,663 @@ static const double power_of_ten[] = {
-    0x4cdc331d57fa5441, 0xe0133fe4adf8e952,
-    0x58180fddd97723a6, 0x570f09eaa7ea7648,};
-
+const uint64_t power_of_five_128[]= {
```
My reading must have been sloppy somehow. Given that we switched from 10^x to 5^x, where in the code do we add the other 2^x back in?

lemire replied:
The 2^x is not part of the binary significand.

We are mapping w * 10 ^q to m * 2^p but w * 10 ^q = w * 5^q * 2^q. The 2^q is matched by the  2^p.

I will add a remark in the code. Something like this:


/**
 * When mapping numbers from decimal to binary,
 * we go from w * 10^q to m * 2^p but we have
 * 10^q = 5^q * 2^q, so effectively
 * we are trying to match
 * w * 2^q * 5^q to m * 2^p. Thus the powers of two
 * are not a concern since they can be represented
 * exactly using the binary notation, only the powers of five
 * affect the binary significand.
 */

jkeiser reviewed (approved):
I have some questions, and I like the factoring to 5^x. Generally better comments, and it's a little simpler to boot.

lemire commented:
@jkeiser I will address all of your comments.

lemire commented:
@jkeiser Can you help me with merging with the main branch? I don't mind doing it, but to avoid issues, I'd like to get some guidance.

Note that I have improved this PR but fine tuning the algorithm quite a bit. If you review the code, you will see that we are much less likely to fall back.

lemire commented:
@jkeiser I think I have addressed all of your comments but I still need to sync with the main branch. As already stated, you might get a kick out of the latest changes... this is getting much fancier.

lemire commented:
I cannot pull more commits into this pull request, so I am closing it in favour of https://github.com/simdjson/simdjson/pull/1222
